week7/answers.sql

```diff
@@ -1,3 +1,263 @@
+        FOREIGN KEY (character_id)
+        REFERENCES characters (character_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE,
+    CONSTRAINT equipped_fk_items
+        FOREIGN KEY (item_id)
+        REFERENCES items (item_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE
+);
+
+CREATE OR REPLACE VIEW character_items AS
+	SELECT c.character_id, c.name AS character_name, it.name AS item_name, it.armor, it.damage
+		FROM characters c
+			INNER JOIN inventory i
+				ON c.character_id = i.character_id
+			INNER JOIN items it
+				ON i.item_id = it.item_id
+	UNION
+	SELECT c.character_id, c.name AS character_name, it.name AS item_name, it.armor, it.damage
+		FROM characters c
+			INNER JOIN equipped e
+				ON c.character_id = e.character_id
+			INNER JOIN items it
+				ON e.item_id = it.item_id
+	ORDER BY item_name ASC;
+
+CREATE OR REPLACE VIEW team_items AS
+	SELECT t.team_id, t.name AS team_name, it.name AS item_name, it.armor, it.damage
+		FROM teams t
+			INNER JOIN team_members tm
+				ON t.team_id = tm.team_id
+			INNER JOIN inventory i
+				ON tm.character_id = i.character_id
+			INNER JOIN items it
+				ON i.item_id = it.item_id
+	UNION
+	SELECT t.team_id, t.name AS team_name, it.name AS item_name, it.armor, it.damage
+		FROM teams t
+			INNER JOIN team_members tm
+				ON t.team_id = tm.team_id
+			INNER JOIN equipped e
+				ON tm.character_id = e.character_id
+			INNER JOIN items it
+				ON e.item_id = it.item_id
+	ORDER BY item_name ASC;
+
+DELIMITER ;;
+
+CREATE FUNCTION armor_total(character_id INT UNSIGNED)
+RETURNS INT UNSIGNED
+DETERMINISTIC
+BEGIN
+	DECLARE gear_ac INT UNSIGNED;
+	DECLARE natural_ac INT UNSIGNED;
+	
+	SELECT SUM(it.armor) INTO gear_ac
+		FROM items it
+			INNER JOIN equipped e
+				ON it.item_id = e.item_id
+		WHERE character_id = e.character_id;
+	
+	SELECT cs.armor INTO natural_ac
+		FROM character_stats cs
+		WHERE character_id = cs.character_id;
+	
+	RETURN gear_ac + natural_ac;
+END ;;
+
+CREATE PROCEDURE attack(defender_id INT UNSIGNED, weapon_id INT UNSIGNED)
+BEGIN
+    DECLARE armor INT UNSIGNED;
+    DECLARE damage INT UNSIGNED;
+    DECLARE result INT SIGNED;
+    DECLARE hp INT SIGNED;
+    
+    SELECT armor_total(defender_id) INTO armor;
+    
+    SELECT it.damage INTO damage
+		FROM equipped e
+			INNER JOIN items it
+				ON e.item_id = it.item_id
+		WHERE e.equipped_id = weapon_id;
+        
+    SET result = damage - armor;
+    
+    SELECT cs.health INTO hp
+		FROM character_stats cs
+        	WHERE defender_id = cs.character_id;
+    
+    IF result > 0 THEN
+		SET hp = hp - result;
+			IF hp <= 0 THEN
+				DELETE FROM characters WHERE character_id = defender_id;
+			END IF;
+		UPDATE character_stats SET health = hp WHERE character_id = defender_id;
+    END IF;
+    
+END ;;
+
+CREATE PROCEDURE equip(equip_id INT UNSIGNED)
+BEGIN
+    DECLARE char_id INT UNSIGNED;
+    DECLARE item INT UNSIGNED;
+    
+    SELECT i.character_id INTO char_id
```

Comment:
These two SELECTs can be merged into a single SELECT:

```sql
SELECT character_id, item_id INTO char_id, item
    FROM inventory
    WHERE equip_id = inventory_id;
```

(No issues with the table alias; I've just copy-pasted the above code a few times.)